Fix pmm deadlock

sparchatus opened the pull request:
An ugly first attempt to allow pmm frame array refill without deadlocking.

Intended to fix #343

sparchatus commented:
Optimally, we would refill the 4K frames while still holding the paging lock. However, this would need a new interface in pmm to allow refill without taking the paging lock again.

sparchatus commented:
there is still the deadlock issue where paging wants to get a frame but pmm is waiting for paging to map a new array. It looks like some kind of backoff is required.

sparchatus commented:
based on #344

Also fixes a (potential?) bug where `total_free_frames` did not count the early array frames.

jcjgraf reviewed (dismissed):
I am no longer able to reproduce the deadlock 👍

wipawel commented on mm/pmm.c:
```diff
@@ -41,11 +41,28 @@ static frames_array_t early_frames;
+static void try_create_4k_frames(void);
+
+static void spin_lock_norefill() {
+    spin_lock(&lock);
+
+    while (refilling) {
+        spin_unlock(&lock);
+        /* TODO: some backoff would be nice here... */
+        spin_lock(&lock);
+    }
```
This code is unsettling. Feels like an antipattern to me. 

You correctly identified the problem being with roundtrips between VMM and PMM due to allocation of new array frames for PMM. Best would be to get rid of it entirely, but we want to preserve the dynamic nature of PMM frames allocation. Therefore, owner of the locks (the initiator of the operation) needs to be able to finish with all the roundtrips, while new operations (by other CPUs) wait on the locks. Indeed, looks like a job for a mutex of some sort.

But maybe (I have not had time to look closely into this) it is possible to add non-locking internal interfaces to both VMM and PMM and use them for the problematic roundtrips. Assumption to check would be if the locks already held at the start of the operation are sufficient protection against other CPUs.

sparchatus replied:
Yeah this one is horrible :sweat_smile: . I now try to solve it using two-stage locking instead.

wipawel commented on mm/pmm.c:
```diff
@@ -674,3 +712,35 @@ void map_frames_array(void) {
+    spin_lock(&priority_lock);
+
+    /* if a refill is already active then they are responsible */
+    if (!refilling) {
```
maybe with `refilling` variable made atomic, this check can be done before taking the lock?

sparchatus replied:
I think we could do it before and after taking the lock but I don't think it is sufficient to only check before.
```
Core 0: pmm_lock()             -> ...              -> refilling = true; unlock(priority) -> ...
Core 1: refill_from_paging()   -> if(!refilling)   -> ...                                -> lock(priority) 
```
Here paging would think there is no refilling but there is.

wipawel commented on mm/pmm.c:
```diff
@@ -108,7 +128,19 @@ static frames_array_t *new_frames_array(void) {
-        array = vmap_kern_4k(mfn_to_virt_map(frame->mfn), frame->mfn, L1_PROT);
+        /* switch to special refilling mode to avoid deadlock with paging */
+        ASSERT(!refilling);
+        refilling = true;
```
does it need to be handled under the lock below?

sparchatus replied:
I think it [does](https://github.com/KernelTestFramework/ktf/pull/345#discussion_r1899418303).

wipawel commented on mm/pmm.c:
```diff
@@ -108,7 +128,19 @@ static frames_array_t *new_frames_array(void) {
+
+        /* switch back to normal mode */
+        ASSERT(refilling);
+        refilling = false;
```
same question as above

sparchatus replied:
I think it [does](https://github.com/KernelTestFramework/ktf/pull/345#discussion_r1899418303).

wipawel commented on mm/pmm.c:
```diff
@@ -95,9 +114,10 @@ static inline void init_frames_array(frames_array_t *array) {
     for (unsigned i = 0; i < ARRAY_SIZE(array->frames); i++)
         init_frame(&array->frames[i]);
     list_add(&array->list, &frames);
+    total_free_frames += array->meta.free_count;
```
good idea of moving it here

wipawel commented on mm/pmm.c:
```diff
@@ -117,13 +149,19 @@ static frames_array_t *new_frames_array(void) {
 
-    total_free_frames += array->meta.free_count;
+    /* since paging will back off from refilling we are responsible to do it here */
+    try_create_4k_frames();
```
hmm... why do we need to split frames during a frame array page mapping?

sparchatus replied:
In this approach, if paging detects that refilling is already active, it will not recreate the minimum 4k frames itself but delegate it to whoever is already refilling. This avoids double refilling of array frames and also avoids inconsistencies due to two interleaved array refills.

sparchatus replied:
Just realized that with this approach we can run out of frames. We cannot guarantee that the thread doing the refill will be the next to acquire the paging lock. If a *lot* of paging is going on this might consume all the reserved frames before the refill gets hold of its page. Hence, I removed the delegation. This might lead to two array frames being allocated when one would have been sufficient.

wipawel commented on mm/pmm.c:
```diff
@@ -117,13 +149,19 @@ static frames_array_t *new_frames_array(void) {
     UNREACHABLE();
 }
 
+static frames_array_t *new_frames_array(void) {
```
nit: could be inline

wipawel commented on mm/pmm.c:
```diff
@@ -41,11 +41,30 @@ static frames_array_t early_frames;
+
+static void try_create_4k_frames(void);
+
+static void spin_lock_standard() {
```
nit: I would rename it to `pmm_lock()` or something like that

wipawel commented on arch/x86/pagetables.c:
```diff
@@ -420,6 +421,26 @@ void *vmap_4k(cr3_t *cr3_ptr, void *va, mfn_t mfn, unsigned long l1_flags,
     return va;
 }
 
+void *vmap_frame_refill(void *va, mfn_t mfn, bool paging_lock) {
+    unsigned long _va = _ul(va) & PAGE_ORDER_TO_MASK(PAGE_ORDER_4K);
+
+    dprintk("%s: va: 0x%p mfn: 0x%lx\n", __func__, va, mfn);
+
+    if (paging_lock)
+        spin_lock(&vmap_lock);
+    else
+        ASSERT(vmap_lock);
+
+    va = _vmap_4k(&cr3, _ptr(_va), mfn, L1_PROT, false);
+
+    if (paging_lock)
+        spin_unlock(&vmap_lock);
+    else
+        ASSERT(vmap_lock);
+
+    return va;
+}
+
```
nit: instead of the `paging_lock` check I would create two versions of this function: one with locking and another with the ASSERT (or maybe a BUG_ON). Despite code duplication, it probably will pay off later when debugging something else.

wipawel commented on mm/pmm.c:
```diff
@@ -108,7 +128,19 @@ static frames_array_t *new_frames_array(void) {
     if (!boot_flags.virt)
         array = (frames_array_t *) mfn_to_virt_kern(frame->mfn);
     else {
-        array = vmap_kern_4k(mfn_to_virt_map(frame->mfn), frame->mfn, L1_PROT);
+        /* switch to special refilling mode to avoid deadlock with paging */
+        ASSERT(!refilling);
+        refilling = true;
+
+        /* only paging will be allowed to take the lock for writing while refilling */
+        spin_unlock(&priority_lock);
+        array = vmap_frame_refill(mfn_to_virt_map(frame->mfn), frame->mfn, from_paging);
+        spin_lock(&priority_lock);
+
+        /* switch back to normal mode */
+        ASSERT(refilling);
+        refilling = false;
```
it might make sense to handle all this code as independent functions (see my comment for the `vmap_frame_refill`).

wipawel commented on mm/pmm.c:
```diff
@@ -41,11 +41,30 @@ static frames_array_t early_frames;
 
 static spinlock_t lock = SPINLOCK_INIT;
 
+static bool refilling;
```
this could be atomic variable, maybe?

sparchatus replied:
I removed it entirely.

sparchatus committed:
pmm: fix total_free_frames initial count

Signed-off-by: Sandro Rüegge <rueegges@ethz.ch>

sparchatus committed:
pagetables,pmm: fix deadlock

1. paging gets exclusive pmm access during array refill.
2. pmm can call back into paging while holding the paging lock.
3. paging performs pmm refill if needed to ensure reserved frames.

Signed-off-by: Sandro Rüegge <rueegges@ethz.ch>

wipawel commented on arch/x86/pagetables.c:
```diff
@@ -420,6 +421,18 @@ void *vmap_4k(cr3_t *cr3_ptr, void *va, mfn_t mfn, unsigned long l1_flags,
+    dprintk("%s: va: 0x%p mfn: 0x%lx\n", __func__, va, mfn);
+
+    BUG_ON(!vmap_lock);
+    va = _vmap_4k(&cr3, _ptr(_va), mfn, l1_flags, false);
```
nit: please add a comment about directly referenced CR3 variable (this will be subject of a rework when we finally get to the better handling of pagetables)

sparchatus replied:
I change the function to match `vmap_kern_4k`/`vmap_kern_4k` with `vmap_4k_nolock`/`vmap_kern_4k_nolock` such that I reference cr3 like the other paging functions

wipawel commented on arch/x86/pagetables.c:
```diff
@@ -420,6 +421,18 @@ void *vmap_4k(cr3_t *cr3_ptr, void *va, mfn_t mfn, unsigned long l1_flags,
+
+    BUG_ON(!vmap_lock);
+    va = _vmap_4k(&cr3, _ptr(_va), mfn, l1_flags, false);
+    BUG_ON(!vmap_lock);
```
Maybe these BUG_ONs should be asserts instead? Please also add a comment why they are there.

wipawel commented on mm/pmm.c:
```diff
@@ -41,10 +41,42 @@ static frames_array_t early_frames;
+
+static void try_create_4k_frames(void);
+
+static void pmm_lock() {
```
nit: please add `inline`

wipawel commented on mm/pmm.c:
```diff
@@ -41,10 +41,42 @@ static frames_array_t early_frames;
+    spin_lock(&priority_lock);
+}
+
+static void pmm_unlock() {
```
nit: please add `inline`

wipawel commented on mm/pmm.c:
```diff
@@ -108,7 +141,49 @@ static frames_array_t *new_frames_array(void) {
+    UNREACHABLE();
+}
+
+static frames_array_t *new_frames_array_nolock(void) {
+    frames_array_t *array;
+    frame_t *frame;
+
+    frame = reserve_frame(get_first_frame(free_frames, PAGE_ORDER_4K));
+    if (!frame)
+        goto error;
+
+    if (!boot_flags.virt)
+        array = (frames_array_t *) mfn_to_virt_kern(frame->mfn);
+    else {
+        /* switch to special refilling mode to avoid deadlock with paging */
+        spin_unlock(&priority_lock);
+
+        /* only paging is allowed to use pmm with only the priority_lock */
+        array = vmap_pmm_refill_nolock(mfn_to_virt_map(frame->mfn), frame->mfn, L1_PROT);
```
since this function is almost identical to the previous one (modulo the vmap callee, maybe define it above the previous one as a helper function (i.e. prefixed with an `_`) and a parameter specifying the callee. Then you can derive two small inline functions from it: the `new_frames_array` and the `new_frames_array_nolock`. Check the vmap_kern_4k(), it might be an inline function, so pass a pointer to the real deal.

sparchatus replied:
I can merge them again but the split was made because of this: https://github.com/KernelTestFramework/ktf/pull/345#discussion_r1899396681

wipawel replied:
Yes, I had something else in mind. Just did not specify it back then. Sorry.

wipawel commented on mm/pmm.c:
```diff
@@ -41,10 +41,42 @@ static frames_array_t early_frames;
+/* avoid recursive refilling after paging */
+static bool refill_from_paging_ongoing = false;
+
+static void try_create_4k_frames(void);
```
nit: it is needed?

sparchatus replied:
Without it, paging will recursively call refill because during refill it calls paging which then again would call refill...

wipawel replied:
I was merely asking if the prototype definition is needed

wipawel commented on mm/pmm.c:
```diff
@@ -674,3 +748,41 @@ void map_frames_array(void) {
+    spin_lock(&priority_lock);
+
+    /* avoid refill_from_paging being called from refill_from_paging */
+    if (!refill_from_paging_ongoing) {
```
since this is the only scope where this variable is being used, declare it here as a `static` variable.

wipawel replied:
Also, I think this variable should be declared as an atomic one and handled via atomic functions

sparchatus replied:
I just realized that since the caller of this function has to hold the vmap_lock we can move this check outside of the spin lock and don't need to use atomics either.
